src/ui/public/modules.js

```diff
@@ -1,9 +1,43 @@
+/**
+ * This module is used by Kibana to create and reuse angular modules. Angular modules
+ * can only be created once and need to have their dependencies at creation. This is
+ * hard/impossible to do in require.js since all of the dependencies for a module are
+ * loaded before it is.
+ *
+ * Before taking this approach we saw three possible solutions:
+ *  1. replicate our js modules in angular modules/use a different module per file
+ *  2. create a single module outside of our js modules and share it
+ *  3. use a helper lib to dynamically create modules as needed.
+ *
+ * We decided to go with #3
+ *
+ * This ends up working by creating a list of modules that the code base creates by
+ * calling `modules.get(name)` with different names, and then before bootstrapping
+ * the application kibana uses `modules.link()` to set the dependencies of the "kibana"
+ * module to include every defined module. This guarantees that kibana can always find
+ * any angular dependecy defined in the kibana code base. This **also** means that
+ * Private modules are able to find any dependency, since they are injected using the
+ * "kibana" module's injector.
+ *
+ * RECOMMENDATIONS:
+ *  - just use the "kibana" module for everything
+ *  - rely on js modules for modularity
+ *  - use Private modules for everything except directives and filters
+ *
```

Comment:
These recommendations should probably also go in the style guide https://github.com/elastic/kibana/blob/master/STYLEGUIDE.md#defining-modules